### PR TITLE
Variable provider updates

### PR DESCRIPTION
**Pull request by JakeGinnivan**
Pre-release tag calculated as unstable.4, and 5 commits since last tag

### Continuous Delivery
"SemVer":"1.2.3-unstable.4"  
"SemVer":"1.2.3",

### Continuous Deployment
"SemVer":"1.2.3-unstable.5"  
"SemVer":"1.2.3.5"

I can't think of anything else to do when the version is stable?

### CODE REVIEW

**Comment by JakeGinnivan** [comment]
// @SimonCropp @andreasohlund @yannisgu @gep13 

Have a look at the approval tests for all the variables

File: GitVersionCore/OutputVariables/VariableProvider.cs [hunk 1]
```diff
@@ -1,71 +1,47 @@
...
-                {BranchName, bmd.Branch},
-                {Sha, bmd.Sha},
-            };
+                // For continuous deployment the commits since tag gets promoted to the pre-release number
+                semanticVersion = new SemanticVersion(semanticVersion);
+                if (semanticVersion.PreReleaseTag.HasTag())
+                {
+                    var oldPreReleaseNumber = semanticVersion.PreReleaseTag.Number;
+                    semanticVersion.PreReleaseTag.Number = semanticVersion.BuildMetaData.CommitsSinceTag;
+                    semanticVersion.BuildMetaData.CommitsSinceTag = oldPreReleaseNumber;
```
Comment (yannisgu) [inline GitVersionCore/OutputVariables/VariableProvider.cs]: Nice idea of inverting PreReleaseTag and BuildMetaData numbers!

File: GitVersionCore/OutputVariables/VariableProvider.cs [hunk 2]
```diff
@@ -1,71 +1,47 @@
...
+                }
+                else
+                {
+                    // When there is no pre-release tag we will make the version a 4 part number
+                    formatProvider = new CommitsAsFourthVersionPartFormatter();
```
Comment (yannisgu) [inline GitVersionCore/OutputVariables/VariableProvider.cs]: I would only do this if `semanticVersion.BuildMetaData.CommitsSinceTag` is not 0. Version like `1.2.3.0` seems wrong to me.

Comment (JakeGinnivan) [inline GitVersionCore/OutputVariables/VariableProvider.cs]: Added this as a discussion point on the main PR

**Pull request merged**